### DISCUSSION TITLE
[knitr] Code Externalization: Use an external R script with your document | /knitr/demo/externalization/

### DISCUSSION THREAD

**Discussion by yihui** [rbind/yihui#765] [Comments]
https://yihui.org/knitr/demo/externalization/

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
There is no read_source(); there is only read_chunk(). This function only reads the code; it does not evaluate the code as source() does.

> Originally posted on 2012-06-04 14:25:13

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
It is language agnostic, meaning you can use any format including LaTeX, HTML, Markdown, ...

writeLines(readLines("path-to-r-script.R")) only reads and writes the code; it is not really read into knitr -- you have to use read_chunk() to read code chunks into knitr and reference them later.

> Originally posted on 2012-07-13 19:20:49

**Reply by yihui** [reply]
That is the pandoc syntax. The original markdown syntax does not support that.

> Originally posted on 2012-07-14 17:36:06

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
I just tested a python script and it worked.

```{r}  
read_chunk('[test.py](http://test.py)')  
```

```{r abc, engine='python'}  
```

In [test.py](http://test.py) I have

## @knitr abc  
print(1)

> Originally posted on 2012-09-24 05:20:47

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Did you see the links to Rnw source files?

But isn't that the default style in knitr? I did use tikz with external=TRUE. The syntax highlighting is a little bit different now, though.

> Originally posted on 2012-11-30 07:17:30

**Reply by yihui** [reply]
I'm not really sure exactly what you want -- a minimal example? <https://github.com/yihui/knitr-examples/blob/master/059-tikz.Rnw>

> Originally posted on 2012-11-30 16:32:14

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
That error dialog box is not helpful; see this manual on how to debug Sweave or knitr in LyX: <https://github.com/downloads/yihui/lyx/sweave.pdf>

> Originally posted on 2012-12-10 06:19:59

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Can you paste the full message log please? e.g. use [pastebin.com](http://pastebin.com)

> Originally posted on 2013-01-08 22:40:40

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
library(knitr)

> Originally posted on 2013-01-08 06:14:31

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
1. I do not know what is your desired output. I use GIT to check if my output has changed since last run, e.g. <https://github.com/yihui/knitr-examples> In this case I only need to guarantee my first run generates the desired output.

2. see the purl() function (to extract R script from *.Rrst)

> Originally posted on 2013-01-12 19:15:54

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
.. {r setup, echo=FALSE}  
opts_chunk$set(echo = FALSE)  
..

> Originally posted on 2013-01-12 19:16:41

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
What error did you see? I just ran the two examples and I did not find any problems.

> Originally posted on 2013-05-25 07:14:13

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
I do not quite recommend you to put too much code into the document. I think it is legitimate to simply source() the script. There is no need to read_chunk().

> Originally posted on 2013-08-22 04:19:21

[a comment by an automated account is not shown]

**Comment by bryce-carson** [comment]
Is there a reason you use the noweb chunk definition syntax in your final example of this page? You also utilize some knitr chunk options in the name of the chunk. Are you using some fancy AWK scripts to allow yourself to write in a more traditional noweb syntax entirely, then translating it all into a knitr input document?

```
<<external-code, cache=FALSE>>=
read_chunk('foo-bar.R')
@
```
